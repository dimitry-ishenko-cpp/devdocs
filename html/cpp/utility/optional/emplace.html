    <h1 id="firstHeading" class="firstHeading">std::optional&lt;T&gt;::emplace</h1>            <table class="t-dcl-begin">  <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(1)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td><pre data-language="cpp">template&lt; class... Args &gt;
T&amp; emplace( Args&amp;&amp;... args );</pre></td>  <td>
<span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span>
</td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template&lt; class... Args &gt;
constexpr T&amp; emplace( Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(2)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td><pre data-language="cpp">template&lt; class U, class... Args &gt;
T&amp; emplace( std::initializer_list&lt;U&gt; ilist, Args&amp;&amp;... args );</pre></td>  <td>
<span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span>
</td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template&lt; class U, class... Args &gt;
constexpr T&amp; emplace( std::initializer_list&lt;U&gt; ilist, Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   </table> <p>Constructs the contained value in-place. If <code>*this</code> already contains a value before the call, the contained value is destroyed by calling its destructor.</p>
<div class="t-li1">
<span class="t-li">1)</span> Initializes the contained value by <a href="../../language/direct_initialization.html" title="cpp/language/direct initialization">direct-initializing</a> (but not direct-list-initializing) with <code><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code> as parameters.</div> <div class="t-li1">
<span class="t-li">2)</span> Initializes the contained value by calling its constructor with <code>ilist, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code> as parameters. This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw527">std::<span class="me2">is_constructible</span></span></a><span class="sy1">&lt;</span>T, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args<span class="sy3">&amp;&amp;</span>...<span class="sy1">&gt;</span><span class="sy4">::</span><span class="me2">value</span></code> is <code>true</code>.</div>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> args... </td> <td> - </td> <td> the arguments to pass to the constructor </td>
</tr> <tr class="t-par"> <td> ilist </td> <td> - </td> <td> the initializer list to pass to the constructor </td>
</tr> <tr class="t-par-hitem"> <td colspan="3"> Type requirements </td>
</tr> <tr class="t-par-req"> <td colspan="3"> -<code>T</code> must be constructible from <code>Args...</code> for overload <span class="t-v">(1)</span> </td>
</tr> <tr class="t-par-req"> <td colspan="3"> -<code>T</code> must be constructible from <code><a href="../initializer_list.html" title="cpp/utility/initializer list">std::initializer_list</a></code> and <code>Args...</code> for overload <span class="t-v">(2)</span> </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>A reference to the new contained value.</p>
<h3 id="Exceptions"> Exceptions</h3> <p>Any exception thrown by the selected constructor of <code>T</code>. If an exception is thrown, <code>*this</code> does not contain a value after this call (the previously contained value, if any, had been destroyed).</p>
<h3 id="Example"> Example</h3> <div class="t-example"> <div class="cpp source-cpp"><pre data-language="cpp">#include &lt;optional&gt;
#include &lt;iostream&gt;
 
struct A {
    std::string s;
    A(std::string str) : s(std::move(str)), id{n++} { note("+ constructed"); }
    ~A() { note("~ destructed"); }
    A(const A&amp; o) : s(o.s), id{n++} { note("+ copy constructed"); }
    A(A&amp;&amp; o) : s(std::move(o.s)), id{n++} { note("+ move constructed"); }
    A&amp; operator=(const A&amp; other) {
        s = other.s;
        note("= copy assigned");
        return *this;
    }
    A&amp; operator=(A&amp;&amp; other) {
        s = std::move(other.s);
        note("= move assigned");
        return *this;
    }
    inline static int n{};
    int id{};
    void note(auto s) { std::cout &lt;&lt; "  " &lt;&lt; s &lt;&lt; " #" &lt;&lt; id &lt;&lt; '\n'; }
};
 
int main()
{
    std::optional&lt;A&gt; opt;
 
    std::cout &lt;&lt; "Assign:\n";
    opt = A("Lorem ipsum dolor sit amet, consectetur adipiscing elit nec.");
 
    std::cout &lt;&lt; "Emplace:\n";
    // As opt contains a value it will also destroy that value
    opt.emplace("Lorem ipsum dolor sit amet, consectetur efficitur.");
 
    std::cout &lt;&lt; "End example\n";
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">Assign:
  + constructed #0
  + move constructed #1
  ~ destructed #0
Emplace:
  ~ destructed #1
  + constructed #2
End example
  ~ destructed #2</pre></div> </div> <h3 id="Defect_reports"> Defect reports</h3> <p>The following behavior-changing defect reports were applied retroactively to previously published C++ standards.</p>
<table class="dsctable"> <tr> <th>DR </th> <th>Applied to </th> <th>Behavior as published </th> <th>Correct behavior </th>
</tr> <tr> <td>
<a rel="nofollow" class="external text" href="https://wg21.link/P2231R1">P2231R1</a> </td> <td>C++20 </td> <td>
<code>emplace</code> was not constexpr while the required operations can be constexpr in C++20 </td> <td>made constexpr </td>
</tr>
</table> <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="operator=.html" title="cpp/utility/optional/operator="> <span class="t-lines"><span>operator=</span></span></a></div> </td> <td> assigns contents <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/optional/emplace" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/optional/emplace</a>
  </p>
</div>
