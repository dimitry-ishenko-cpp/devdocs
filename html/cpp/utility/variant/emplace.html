    <h1 id="firstHeading" class="firstHeading">std::variant&lt;Types...&gt;::emplace</h1>            <table class="t-dcl-begin">  <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(1)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td><pre data-language="cpp">template &lt;class T, class... Args&gt;
T&amp; emplace(Args&amp;&amp;... args);</pre></td>  <td>
<span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span>
</td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template &lt;class T, class... Args&gt;
constexpr T&amp; emplace(Args&amp;&amp;... args);</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(2)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td><pre data-language="cpp">template &lt;class T, class U, class... Args&gt;
T&amp; emplace( std::initializer_list&lt;U&gt; il, Args&amp;&amp;... args );</pre></td>  <td>
<span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span>
</td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template &lt;class T, class U, class... Args&gt;
constexpr T&amp; emplace( std::initializer_list&lt;U&gt; il, Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(3)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td><pre data-language="cpp">template &lt;std::size_t I, class... Args&gt;
std::variant_alternative_t&lt;I, variant&gt;&amp; emplace( Args&amp;&amp;... args );</pre></td>  <td>
<span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span>
</td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template &lt;std::size_t I, class... Args&gt;
constexpr std::variant_alternative_t&lt;I, variant&gt;&amp; emplace( Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(4)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td><pre data-language="cpp">template &lt;std::size_t I, class U, class... Args&gt;
std::variant_alternative_t&lt;I, variant&gt;&amp; 
    emplace( std::initializer_list&lt;U&gt; il, Args&amp;&amp;... args );</pre></td>  <td>
<span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span>
</td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template &lt;std::size_t I, class U, class... Args&gt;
constexpr std::variant_alternative_t&lt;I, variant&gt;&amp; 
    emplace( std::initializer_list&lt;U&gt; il, Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   </table> <p>Creates a new value in-place, in an existing <code>variant</code> object</p>
<span class="t-li">1)</span> Equivalent to <code>emplace<span class="sy1">&lt;</span>I<span class="sy1">&gt;</span><span class="br0">(</span><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...<span class="br0">)</span></code>, where <code>I</code> is the zero-based index of <code>T</code> in <code>Types...</code>. <ul><li> This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw528">std::<span class="me2">is_constructible_v</span></span></a><span class="sy1">&lt;</span>T, Args...<span class="sy1">&gt;</span></code> is <code>true</code>, and <code>T</code> occurs exactly once in <code>Types...</code>
</li></ul> <span class="t-li">2)</span> Equivalent to <code>emplace<span class="sy1">&lt;</span>I<span class="sy1">&gt;</span><span class="br0">(</span>il, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...<span class="br0">)</span></code>, where <code>I</code> is the zero-based index of <code>T</code> in <code>Types...</code>. <ul><li> This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw528">std::<span class="me2">is_constructible_v</span></span></a><span class="sy1">&lt;</span>T, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args...<span class="sy1">&gt;</span></code> is <code>true</code>, and <code>T</code> occurs exactly once in <code>Types...</code>
</li></ul> <span class="t-li">3)</span> First, destroys the currently contained value (if any). Then <a href="../../language/direct_initialization.html" title="cpp/language/direct initialization">direct-initializes</a> the contained value as if constructing a value of type <code>T_I</code> with the arguments <code><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>. If an exception is thrown, <code>*this</code> may become <a href="valueless_by_exception.html" title="cpp/utility/variant/valueless by exception">valueless_by_exception</a>. <ul>
<li> This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw528">std::<span class="me2">is_constructible_v</span></span></a><span class="sy1">&lt;</span>T_I, Args...<span class="sy1">&gt;</span></code> is <code>true</code>. </li>
<li> It is a compile-time error if <code>I</code> is not less than <code>sizeof...(Types)</code>.</li>
</ul> <span class="t-li">4)</span> First, destroys the currently contained value (if any). Then <a href="../../language/direct_initialization.html" title="cpp/language/direct initialization">direct-initializes</a> the contained value as if constructing a value of type <code>T_I</code> with the arguments <code>il, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>. If an exception is thrown, <code>*this</code> may become <a href="valueless_by_exception.html" title="cpp/utility/variant/valueless by exception">valueless_by_exception</a>. <ul>
<li> This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw528">std::<span class="me2">is_constructible_v</span></span></a><span class="sy1">&lt;</span>T_I, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args...<span class="sy1">&gt;</span></code> is <code>true</code>. </li>
<li> It is a compile-time error if <code>I</code> is not less than <code>sizeof...(Types)</code>.</li>
</ul>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> args </td> <td> - </td> <td> constructor arguments to use when constructing the new value </td>
</tr> <tr class="t-par"> <td> il </td> <td> - </td> <td> initializer_list argument to use when constructing the new value </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>A reference to the new contained value.</p>
<h3 id="Exceptions"> Exceptions</h3> <div class="t-li1">
<span class="t-li">1-4)</span> Any exception thrown during the initialization of the contained value.</div> <h3 id="Example"> Example</h3> <div class="t-example"> <div class="cpp source-cpp"><pre data-language="cpp">#include &lt;iostream&gt;
#include &lt;string&gt;
#include &lt;variant&gt;
 
int main()
{
    std::variant&lt;std::string&gt; v1;
    v1.emplace&lt;0&gt;("abc"); // OK
    std::cout &lt;&lt; std::get&lt;0&gt;(v1) &lt;&lt; '\n';
    v1.emplace&lt;std::string&gt;("def"); // OK
    std::cout &lt;&lt; std::get&lt;0&gt;(v1) &lt;&lt; '\n';
 
    std::variant&lt;std::string, std::string&gt; v2;
    v2.emplace&lt;1&gt;("ghi"); // OK
    std::cout &lt;&lt; std::get&lt;1&gt;(v2) &lt;&lt; '\n';
    // v2.emplace&lt;std::string&gt;("abc"); -&gt; Error
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">abc
def
ghi</pre></div> </div> <h3 id="Defect_reports"> Defect reports</h3> <p>The following behavior-changing defect reports were applied retroactively to previously published C++ standards.</p>
<table class="dsctable"> <tr> <th>DR </th> <th>Applied to </th> <th>Behavior as published </th> <th>Correct behavior </th>
</tr> <tr> <td>
<a rel="nofollow" class="external text" href="https://wg21.link/P2231R1">P2231R1</a> </td> <td>C++20 </td> <td>
<code>emplace</code> was not constexpr while the required operations can be constexpr in C++20 </td> <td>made constexpr </td>
</tr>
</table> <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="operator=.html" title="cpp/utility/variant/operator="> <span class="t-lines"><span>operator=</span></span></a></div> </td> <td> assigns a variant <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/variant/emplace" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/variant/emplace</a>
  </p>
</div>
