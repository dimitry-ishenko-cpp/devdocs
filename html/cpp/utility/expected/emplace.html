    <h1 id="firstHeading" class="firstHeading">std::expected&lt;T,E&gt;::emplace</h1>            <table class="t-dcl-begin"> <tr class="t-dcl-h"> <td> T is not <i>cv</i> void</td> <td></td> <td></td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class... Args &gt;
constexpr T&amp; emplace( Args&amp;&amp;... args ) noexcept;</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class U, class... Args &gt;
constexpr T&amp; emplace( std::initializer_list&lt;U&gt;&amp; il, Args&amp;&amp;... args ) noexcept;</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl-h"> <td> T is <i>cv</i> void</td> <td></td> <td></td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">constexpr void emplace() noexcept;</pre>
</td> <td> (3) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <p>Constructs an expected value in-place. After the call, <a href="operator_bool.html" title="cpp/utility/expected/operator bool"><code>has_value()</code></a> returns true.</p>
<div class="t-li1">
<span class="t-li">1)</span> Destroys the contained value, then initializes the expected value contained in <code>*this</code> as if by <a href="../../language/direct_initialization.html" title="cpp/language/direct initialization">direct-initializing</a> an object of type <code>T</code> from the arguments <code><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>. This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw532">std::<span class="me2">is_nothrow_constructible_v</span></span></a><span class="sy1">&lt;</span>T, Args...<span class="sy1">&gt;</span></code> is true.</div> <div class="t-li1">
<span class="t-li">2)</span> Destroys the contained value, then initializes the expected value contained in <code>*this</code> as if by <a href="../../language/direct_initialization.html" title="cpp/language/direct initialization">direct-initializing</a> an object of type <code>T</code> from the arguments <code>il, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1120">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>. This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw532">std::<span class="me2">is_nothrow_constructible_v</span></span></a><span class="sy1">&lt;</span>T, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args...<span class="sy1">&gt;</span></code> is true.</div> <div class="t-li1">
<span class="t-li">3)</span> If <code>*this</code> contains an unexpected value, destroys that value.</div>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> args... </td> <td> - </td> <td> the arguments to pass to the constructor </td>
</tr> <tr class="t-par"> <td> il </td> <td> - </td> <td> the initializer list to pass to the constructor </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>A reference to the new contained value.</p>
<h3 id="Exceptions"> Exceptions</h3> <div class="t-noexcept-inline">
<a href="../../language/noexcept_spec.html" title="cpp/language/noexcept spec"><code>noexcept</code></a> specification: <div class="t-noexcept-box"><span class="t-cc"><code>noexcept</code></span></div> </div> <h3 id="Notes"> Notes</h3> <p>If the construction of <code>T</code> is potentially-throwing, this function is not defined. In this case, it is the responsibility of the user to create a temporary and move or copy it.</p>
<h3 id="Example"> Example</h3>  <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="operator=.html" title="cpp/utility/expected/operator="> <span class="t-lines"><span>operator=</span></span></a></div> </td> <td> assigns contents <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/expected/emplace" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/expected/emplace</a>
  </p>
</div>
